### PR TITLE
rad-profile: the rad-profile CLI library

### PR DESCRIPTION
**Pull request by FintanH**
Currently `ssh_add` is a stub since there is some reworking in `thrussh` required to get https://github.com/radicle-dev/radicle-keystore/pull/17 in.

### CODE REVIEW

**Comment by FintanH** [comment]
This is ready for some inspection :female_detective:

File: link-crypto/Cargo.toml [hunk 1]
```diff
@@ -33,6 +32,11 @@ features = ["std", "derive"]
 path = "../git-ext"
 features = ["serde", "minicbor"]
 
+[dependencies.radicle-keystore]
+git = "https://github.com/radicle-dev/radicle-keystore"
+rev = "2a28dd785f7cd360cdc962e4d343fa9f7ac39814"
```
Comment (FintanH) [inline link-crypto/Cargo.toml:37]: Although we should probably wait for merging until a minor release is made[0]

[0]: https://github.com/ZcashFoundation/ed25519-zebra/issues/45#issuecomment-820181394

**Comment by FintanH** [comment]
oh lol, what do we do for Windows? :laughing: `thrussh` has no implementation for `not(unix)` :grimacing:

**Comment by kim** [comment]
>    thrussh has no implementation for not(unix)

Is this because std doesn't export UNIX domain sockets under windows? That's
more a matter of policy it seems [0]. Oh look, a yak!

[0]: https://github.com/rust-lang/rust/issues/56533

**Comment by FintanH** [comment]
> Is this because std doesn't export UNIX domain sockets under windows? That's
more a matter of policy it seems [0]. Oh look, a yak!

Thanks for the pointer :) Well, this would mean that we'd have to also get `tokio` and `smol` to add the Windows support so that in turn it can be supported by `thrussh`. 

image

**Comment by kim** [comment]
>    get tokio and smol to add the Windows support

Yeah right.

I think this is probably the point in time where we no longer pretend we could
support Windows eventually (outside WSL, that is). Or else, someone steps up to
become a maintainer for Windows-compat.

**Comment by FintanH** [comment]
> I think this is probably the point in time where we no longer pretend we could
support Windows eventually (outside WSL, that is). Or else, someone steps up to
become a maintainer for Windows-compat.

I think we can at least punt on this for now with https://github.com/radicle-dev/radicle-link/pull/728/commits/a04a95407a54b0ea4a6120ecc8edf3433abe635a, but ya it doesn't seem feasible with all of us working on unix based machines as well.

The only other option I see is hiding all ssh-agent functionality behind a `cfg(unix)` and forcing windows users to use the `prompt` method of key access :upside_down_face:

File: rad-profile/src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+pub mod args;
+pub mod main;
+
+pub use main::main;
```
Comment (xla) [inline rad-profile/src/cli.rs:9]: How does this interact with the `pub mod main;` above?

File: rad-profile/src/cli/args.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+/// Create a new profile, generating a new secret key and initialising
+/// configurations and storage.
+#[derive(Debug, StructOpt)]
+pub struct Create {}
+
+/// Get the currently active profile.
+#[derive(Debug, StructOpt)]
+pub struct Get {}
```
Comment (xla) [inline rad-profile/src/cli/args.rs:35]: Could this take an optional `id` to get another profile than the currently active?

Comment (xla) [reply rad-profile/src/cli/args.rs:35]: If not it might be more accurately named `GetActive`.

**Review by xla — COMMENTED** [review]
Cool beans! Are we concerned that there are no destructive actions at the moment, e.g. `profile delete` or `ssh remove`?

File: link-crypto/Cargo.toml (hunk 1 above)
Comment (FintanH) [reply link-crypto/Cargo.toml:37]: Would be preferable, I think, for the upstream release to be made. They said they'd make a decision tomorrow :)

File: rad-profile/src/cli.rs (hunk 2 above)
Comment (FintanH) [reply rad-profile/src/cli.rs:9]: I believe Rust knows what's up. I was playing around with a `src/main.rs` locally and I made this change to make it more sensible than `rad_exe::main::main()` :smile:

File: rad-profile/src/cli/args.rs (hunk 3 above)
Comment (FintanH) [reply rad-profile/src/cli/args.rs:35]: Ya, that's a good point. Need to revisit why I didn't do that initially :)

**Comment by FintanH** [comment]
> Cool beans! Are we concerned that there are no destructive actions at the moment, e.g. profile delete or ssh remove?

Very good point. Aside: it's funny `ssh-add` can also delete a key haha. But I don't see why we couldn't add destructive operations.

**Comment by kim** [comment]
>    it's funny ssh-add can also delete a key

From the agent, which is ephemeral storage. Think of it more as "unload key".

I believe it was mentioned in RFC #682 that secure long term storage of key
material is still a concern.

**Comment by FintanH** [comment]
> From the agent, which is ephemeral storage. Think of it more as "unload key".

Ya, good way of putting it. I was just musing about the name of `add` :)

> I believe it was mentioned in RFC #682 that secure long term storage of key
material is still a concern.

Long-term storage as well as rotation I would imagine

**Review by xla — DISMISSED** [review]
🎤 🗑 📏 🛎

**Review by xla — COMMENTED** [review]
This yak looks sufficiently shaved.

⛏ 📘 📙 🎠

**Review by xla — APPROVED** [review]
⚱ 🖨 🏺 🍂

**Pull request merged**